include/picongpu/plugins/Emittance.hpp

```diff
@@ -0,0 +1,1078 @@
+
+                /* remove all non-roots (-1 values) */
+                std::sort( planeReduceRootRanks.begin(), planeReduceRootRanks.end() );
+                std::vector<int> ranks( std::lower_bound( planeReduceRootRanks.begin(),
+                                                          planeReduceRootRanks.end(),
+                                                          0 ),
+                                        planeReduceRootRanks.end() );
+
+                MPI_CHECK(MPI_Comm_group( MPI_COMM_WORLD, &world_group ));
+                MPI_CHECK(MPI_Group_incl( world_group, ranks.size(), ranks.data(), &new_group ));
+                MPI_CHECK(MPI_Comm_create( MPI_COMM_WORLD, new_group, &commGather ));
+                MPI_CHECK(MPI_Group_free( &new_group ));
+                MPI_CHECK(MPI_Group_free( &world_group ));
+            }
+
+            // decide which MPI-rank writes output
+            int gatherRank = -1;
+            if( commGather != MPI_COMM_NULL )
+                MPI_CHECK(MPI_Comm_rank( commGather, &gatherRank ));
+            writeToFile = ( gatherRank == 0 );
+
+            const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+            gSumMom2 = new GridBuffer<
+                float_64,
+                DIM1
+            >( DataSpace< DIM1 >( subGrid.getLocalDomain().size.y() ) );
+            gSumPos2 = new GridBuffer<
+                float_64,
+                DIM1
+            >( DataSpace< DIM1 >( subGrid.getLocalDomain().size.y() ) );
+            gSumMomPos = new GridBuffer<
+                float_64,
+                DIM1
+            >( DataSpace< DIM1 >( subGrid.getLocalDomain().size.y() ) );
+            gCount_e = new GridBuffer<
+                float_64,
+                DIM1
+            >( DataSpace< DIM1 >( subGrid.getLocalDomain().size.y() ) );
+
+            // only MPI rank that writes to file
+            if( writeToFile )
+            {
+                // open output file
+                outFile.open(
+                    filename.c_str( ),
+                    std::ofstream::out | std::ostream::trunc
+                );
+
+                // error handling
+                if( !outFile )
+                {
+                    std::cerr <<
+                        "Can't open file [" <<
+                        filename <<
+                        "] for output, diasble plugin output. " <<
+                        std::endl;
+                    writeToFile = false;
+                }
+            }
+
+            // set how often the plugin should be executed while PIConGPU is running
+            Environment<>::get( ).PluginConnector( ).setNotificationPeriod(
+                this,
+                m_help->notifyPeriod.get( id )
+            );
+        }
+
+        virtual ~CalcEmittance( )
+        {
+            if( writeToFile )
+            {
+                outFile.flush( );
+                // flush cached data to file
+                outFile << std::endl;
```

Comment:
This will add an empty line at the end of the file. This means if you restart the simulation you will have an empty line between the time steps. IMO this must be removed. The cached is already flushed with the command above.